lib/connection/pool.js

```diff
@@ -215,21 +215,30 @@ function stateTransition(self, newState) {
   }
 }
 
-function authenticate(pool, auth, connection, cb) {
-  if (auth[0] === undefined) return cb(null);
+function authenticate(pool, credentials, connection, cb) {
+  // If there are no credentials, do not authenticated
+  if (!credentials) {
+    return cb(null);
+  }
+
+  // HACK: If default mechanism has not been determined yet, do not
+  // try to authenticate
+  if (credentials.mechanism === 'default') {
+    return cb(null);
+  }
+
   // We need to authenticate the server
-  var mechanism = auth[0];
-  var db = auth[1];
+  const mechanism = credentials.mechanism;
   // Validate if the mechanism exists
   if (!pool.authProviders[mechanism]) {
     throw new MongoError(f('authMechanism %s not supported', mechanism));
   }
 
   // Get the provider
-  var provider = pool.authProviders[mechanism];
+  const provider = pool.authProviders[mechanism];
 
   // Authenticate using the provided mechanism
-  provider.auth.apply(provider, [write(pool), [connection], db].concat(auth.slice(2)).concat([cb]));
+  provider.auth.call(provider, write(pool), [connection], credentials, cb);
```

Comment:
Do we really need to `apply` or `call` here? Why can't we just call `provider.auth(...)`?

Reply:
doh 🤦‍♂️ . You're right.